File: services/tu-core/internal/domain/student/models/valueobjects/nisn.go
// Package valueobjects menyediakan value objects untuk domain Student.
// SYMMETRIC dengan frontend (apps/siakad-tu/src/domain/student/models/value-objects/).
package valueobjects

import (
	"errors"
	"fmt"
	"regexp"
)

// NISN adalah value object untuk Nomor Induk Siswa Nasional.
type NISN struct {
	value string
}

var nisnPattern = regexp.MustCompile(`^\d{10}$`)

// NewNISN membuat NISN baru dengan validasi.
func NewNISN(value string) (NISN, error) {
	if value == "" {
		return NISN{}, errors.New("NISN tidak boleh kosong")
	}
	if !nisnPattern.MatchString(value) {
		return NISN{}, fmt.Errorf("NISN harus 10 digit angka, got: %s", value)
	}
	return NISN{value: value}, nil
}

// Value mengembalikan nilai NISN.
func (n NISN) Value() string { return n.value }

// String mengimplementasikan Stringer.
func (n NISN) String() string { return n.value }

// Equals mengecek kesamaan.
func (n NISN) Equals(other NISN) bool { return n.value == other.value }

File: apps/siakad-tu/src/domain/student/value-objects/NISN.js
// apps/siakad-tu/src/domain/student/value-objects/NISN.js

/**
 * Value Object: NISN (Nomor Induk Siswa Nasional)
 *
 * Aturan:
 * - 10 digit angka
 * - Tidak boleh kosong
 * - Immutable setelah dibuat
 * - Self-validating
 */
export class NISN {
  #value;

  constructor(value) {
    this.#validate(value);
    this.#value = value;
    Object.freeze(this);
  }

  #validate(value) {
    if (!value) {
      throw new Error('NISN tidak boleh kosong.');
    }

    const strValue = String(value).trim();

    if (!/^\d{10}$/.test(strValue)) {
      throw new Error(
        `NISN harus terdiri dari 10 digit angka. Diterima: "${value}".`
      );
    }
  }

  get value() {
    return this.#value;
  }

  equals(other) {
    if (!(other instanceof NISN)) return false;
    return this.#value === other.#value;
  }

  toString() {
    return this.#value;
  }

  toJSON() {
    return this.#value;
  }

  static fromJSON(value) {
    return new NISN(value);
  }
}
